### PR TITLE
[RFC] Refactor Neovim to remove the side effects of `K_EVENT`

### LINKED ISSUE #3382: operator-pending mode interrupted by redraw or job/event
Followup of https://github.com/neovim/neovim/pull/3246#issuecomment-136294721 (cannot find an existing issue for this).

To reproduce, use `:term` to run a program that refreshes often (such as `top`):

```
:term top
<c-\><c-n>
<c-w>       <-- operator-pending mode is canceled by the next redraw
```

[Solution](https://github.com/neovim/neovim/pull/3246#issuecomment-136317554) suggested by @tarruda :

> we need to handle K_EVENT in every function that reads another character for the purpose of finishing an operator pending mode. It's not an absurd amount of work, but it requires a search through normal.c and ops.c

### PR DESCRIPTION
**Pull request by tarruda**
This PR is my new attempt to remove the effects of the internal `K_EVENT` key code, which is used to represent events in various modes that explicitly declare the ability to do so. While using a special key to signal events is relatively safe(because Vim is a state machine that only deals with keys), it has a couple of disadvantages:
- Some operators can be interrupted
- Mappings are affected

Previously I have attempted to remove this key by always handling events in the most inner place of Neovim loops: the input.c module, which created instability since we can't handle events in every state.

Fixing this issue is a bigger deal now because terminal redrawing is now handled as deferred events: https://github.com/neovim/neovim/commit/47cbbc0b8c864bdb1aaa572f0396642e2c66deae

Close #3382 (and possibly many other issues I can't find right now)

### CODE REVIEW

File: src/nvim/normal.c [hunk 1]
```diff
@@ -904,164 +877,347 @@ normal_cmd (
...
+  // Postponed until here to avoid computing w_virtcol too often.
+  update_curswant();
 
-  /* Save count before an operator for next time. */
-  opcount = ca.opcount;
+  if (exmode_active) {
+    if (s->noexmode) {
+      return false;
+    }
+    do_exmode(exmode_active == EXMODE_VIM);
```
Comment (oni-link) [inline src/nvim/normal.c]: If `!s->cmdwin || cmdwin_result == 0` is `true` then after the call `do_exmode()`, `normal_cmd()` is now always called. That was not the case in the old code.

File: src/nvim/normal.c [hunk 2]
```diff
@@ -904,164 +877,347 @@ normal_cmd (
...
+      // after insert mode finishes!
+      need_fileinfo = false;
+    }
+  }
+
+  // Reset "got_int" now that we got back to the main loop.  Except when
+  // inside a ":g/pat/cmd" command, then the "got_int" needs to abort
+  // the ":g" command.
+  // For ":g/pat/vi" we reset "got_int" when used once.  When used
+  // a second time we go back to Ex mode and abort the ":g" command. */
```
Comment (oni-link) [inline src/nvim/normal.c]: `*/` can be removed at the end of the line.

File: src/nvim/normal.c [hunk 3]
```diff
@@ -904,164 +877,347 @@ normal_cmd (
...
+  if (got_int) {
+    if (s->noexmode && global_busy && !exmode_active
+        && s->previous_got_int) {
+      // Typed two CTRL-C in a row: go back to ex mode as if "Q" was
+      // used and keep "got_int" set, so that it aborts ":g". */
```
Comment (oni-link) [inline src/nvim/normal.c]: `*/` can be removed at the end of the line.

File: src/nvim/normal.c (hunk 1 above)
Comment (tarruda) [inline src/nvim/normal.c]: :+1:

File: src/nvim/normal.c (hunk 2 above)
Comment (tarruda) [inline src/nvim/normal.c]: :+1:

File: src/nvim/normal.c (hunk 3 above)
Comment (tarruda) [inline src/nvim/normal.c]: :+1:

**Comment by oni-link** [comment]
CursorHold is now retriggered (every `'updatetime'` ms) without pressing a key:

vimrc:
```vim
set updatetime=100

let g:x=0

function! Update()
    let g:x=g:x+1
    echom "CursorHoldI ".g:x
    call jobstart('echo')
endfunction

function! Update2()
    let g:x=g:x+1
    echom "CursorHold ".g:x
    call jobstart('echo') 
endfunction

autocmd CursorHoldI * call Update()
autocmd CursorHold * call Update2()
```
Script used with `nvim -u vimrc`.

**Comment by tarruda** [comment]
> CursorHold is now retriggered (every 'updatetime' ms) without pressing a key:

:+1: Fixed

**Comment by tarruda** [comment]
@ZyX-I I'm having trouble determining why a certain shada test is failing in this PR, more specifically on commit https://github.com/tarruda/neovim/commit/a4be27040537941bf81ce9c552be21ad1bf7616f (main: Refactor main_loop to call `os_inchar` directly).

The failing test can be run with:

```
TEST_FILTER='correctly ignores case' TEST_FILE=test/functional/shada/shada_spec.lua make test
```

and has the following failed assertion:

```
Failure → test/functional/shada/shada_spec.lua @ 189
ShaDa support code correctly ignores case with shada-r option
test/functional/shada/shada_spec.lua:200: Expected objects to be the same.
Passed in:
(table) {
  [7] = 1
  [8] = 2
  [9] = 1
  [10] = 4
 *[11] = 1 }
Expected:
(table) {
  [7] = 1
  [8] = 2
  [9] = 1
  [10] = 4
 *[11] = 2 }
```

That is, key `11` has a expected value of 2 but the actual value was 1. Since I'm having trouble to quickly parse the relevant shada test, can you explain what is the meaning of key `11`?

**Comment by ZyX-I** [comment]
@tarruda This key is a number of recorded changelist items in the ShaDa file for the file given as the argument to `find_file`. Basically this means that one change has disappeared from change list for some reason.

**Comment by bfredl** [comment]
I get this similar error on master
```
ShaDa support code correctly uses shada-r option
/home/bjorn/dev/neovim/test/functional/shada/shada_spec.lua:177: Expected objects to be the same.
Passed in:
(table) { }
Expected:
(table) {
 *[7] = 1
  [8] = 2
  [9] = 1
  [10] = 4
  [11] = 1 }
```

**Comment by ZyX-I** [comment]
@bfredl This is by no means similar error. In your case ShaDa file contains nothing about the tested file *at all*, which is very different from missing changelist entry.

**Comment by tarruda** [comment]
> @tarruda This key is a number of recorded changelist items in the ShaDa file for the file given as the argument to find_file. Basically this means that one change has disappeared from change list for some reason.

I'd like to debug this with gdb, can you point me to the section of code that reads the number of changelist items into the shada msgpack entry?

**Comment by ZyX-I** [comment]
@tarruda This is lua code, function `find_file`.

**Comment by ZyX-I** [comment]
And one ShaDa entry *always* contains one (entry type is 11) or zero (any other entry type) changelist items.

**Comment by tarruda** [comment]
> @tarruda This is lua code, function find_file.

I meant that code that reads the number of changelist items into a msgpack entry(before writing to the shada file). I want to follow the code paths before and after the commit that introduces the failure

**Comment by ZyX-I** [comment]
There is no msgpack entry that contains a number of changelist item. Code that processes change and jump lists is using `MERGE_JUMPS` macros: both when reading and writing the file.

**Comment by tarruda** [comment]
Ok, that is a big macro. Can we turn it into a function so it is easier to debug?

**Comment by tarruda** [comment]
Nvm, I see that you use some defines to change the behavior of the macro, it would not be as simple as converting to a function

**Comment by ZyX-I** [comment]
@tarruda Not in C. There will be a number of `void*` pointers and a number of similarly generic function pointer types.

**Comment by tarruda** [comment]
@ZyX-I nevermind, I just wanted to find the relevant variables and they are `b_changelist`/`b_changelistlen`. This should be enough for me to start debugging

**Comment by tarruda** [comment]
Found the problem with the shada test: Now that K_EVENT doesn't come from `vgetc()`, `may_sync_undo` isn't called for K_EVENT. This is required to set `curbuf->b_new_change` and `curbuf->b_u_synced` which are used to [increment](https://github.com/neovim/neovim/blob/3bb266266935e9bf95f4a776b33800b9303d4607/src/nvim/misc1.c#L2049-L2093) `curbuf->b_changedlistlen`(serialized by shada). Calling `may_sync_undo` before the mode-specific execute function fixed the issue.

**Comment by ZyX-I** [comment]
It is rather strange that this error was caught by a test that exists merely to check that specific file gets ignored with some `shada-r` option value. I.e. it was not testing something related to undo or changelist, just I wrote it so it counts ShaDa entries and not just checks their presence (which would also be enough for the test purposes).

**Comment by tarruda** [comment]
> It is rather strange that this error was caught by a test that exists merely to check that specific file gets ignored with some shada-r option value. I.e. it was not testing something related to undo or changelist, just I wrote it so it counts ShaDa entries and not just checks their presence (which would also be enough for the test purposes).

In a software so complex as this one, it is not unusual for this to happen. When testing feature A, we implicitly test feature B, C, D...

The important is that the shada tests added new coverage for existing code

**Comment by ZyX-I** [comment]
I know. But it would be much better if it was caught using specific test for this feature and not implicit test which may be changed so, that it will still serve its purpose, but not catch additional problems.

**Comment by tarruda** [comment]
I still have a few ideas that will greatly simplify Neovim code, similarly to how the changes implemented by this branch did: Generalizing source code structures. But this will have to wait until after 0.1 because this PR is already too big

I'd really appreciate anyone helping me test this branch by using it locally. I'm relatively confident no major bugs were introduced because we have a reasonable test suite and I've been using it since the PR started, but it would certainly help if more people were using it in different environments.

**Comment by justinmk** [comment]
Will try it for a couple days.

File: src/nvim/README.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+README](https://raw.githubusercontent.com/vim/vim/master/src/README.txt) still
+applies.
+
+This document aims to give a high level overview of how Neovim works internally,
+focusing on parts that are different from Vim. Currently this is still a work in
+progress, especially because I have avoided adding too many details about parts
+that are constantly changing. As the code becomes more organized and stable,
+this document will be updated to reflect the changes.
+
+If you are looking module-specific details, it is best to read the source code.
```
Comment (oni-link) [inline src/nvim/README.md]: ... looking module-... -> ... looking for module-...

Comment (tarruda) [inline src/nvim/README.md]: :+1:

**Comment by justinmk** [comment]
Although this appears to fix https://github.com/neovim/neovim/issues/3382 (by preventing redraw during operator-pending?), this other case still exists:

- `nvim -u NONE -c 'te top'
- `ctrl-\ ctrl-n`
- `:e<tab>` to show wildmenu. Notice that on the next `top` redraw/event, wildmenu disappears.

**Comment by tarruda** [comment]
@justinmk for both problems you reported, I don't think there's much I can do without refactoring screen.c. The real issue is that in some situations, no other information about what is on the screen is kept except for what is already on `ScreenLines` and friends.

For example, when `'wildmenu'` is being shown in command-line mode, screen.c has no idea that it is displaying a wildmenu. It was simply drawn over the status area, which is redraw when `update_screen` is called.

In other words, when you press <kbd>enter</kbd> and `'wildmenu'` is being shown, no function is called to hide the menu and redisplay the status line, it is simply a consequence of `update_screen` resetting status text in the next normal loop iteration.

I suppose I could add an extra global variable that disables status lines redrawing while on command-line, but it could be error prone and cause other bugs. I would rather defer fixing this(which is already present on master) to another PR where screen.c is properly refactored.

**Comment by oni-link** [comment]
The use of `intr_char` inside of a `#ifdef UNIX`was removed in some cases in the refactoring. Do we still need it? Can this global variable be removed?

**Comment by tarruda** [comment]
> The use of intr_char inside of a #ifdef UNIXwas removed in some cases in the refactoring. Do we still need it? Can this global variable be removed?

The global variable is not set and always has a value of 0, so I guess it can be removed safely

**Comment by tarruda** [comment]
In any case, it's not like we want to have platform-specific code in core functions. This translation should happen in the TUI(if it should happen at all)

**Comment by justinmk** [comment]
> The real issue is that in some situations, no other information about what is on the screen is kept except for what is already on ScreenLines and friends.

Understood, though in the case of `nnoremap gl :ls<CR>:buffer<Space>` where the pager is showing with Press-Enter prompt, I believe we do have flags that can already be checked.

For reference, this issue is covered by https://github.com/neovim/neovim/issues/3416

**Pull request merged**